src/tasty/tasty.tsx

```diff
@@ -140,106 +158,163 @@ function tasty<K extends StyleList, C = Record<string, unknown>>(
-      [styleCacheKey],
-    );
+    // eslint-disable-next-line react/display-name
+    _TastyComponent = forwardRef(
+      (allProps: AllBasePropsWithMods<K> & WithVariant<V>, ref) => {
+        const { variant, ...restProps } = allProps;
 
-    let contextBreakpoints = useContext(BreakpointsContext);
+        const Component = (variantComponents[variant as string] ||
+          variantComponents['default']) as ComponentType<
+          AllBasePropsWithMods<K>
+        >;
 
-    breakpoints = breakpoints ?? contextBreakpoints;
+        return <Component {...restProps} ref={ref} />;
+      },
+    );
+  } else {
+    let Element = styledComponents[originalAs](({ css }) => css);
+
+    /**
+     * An additional optimization that allows to avoid rendering styles across various instances
+     * of the same element if no custom styles are provided via `styles` prop or direct style props.
+     */
+    const renderDefaultStyles = cacheWrapper((breakpoints) => {
+      return renderStyles(defaultStyles || {}, pointsToZones(breakpoints));
+    });
 
-    let renderedStyles = useMemo(() => {
-      return useDefaultStyles
-        ? renderDefaultStyles(breakpoints as number[])
-        : renderStyles(allStyles, pointsToZones(breakpoints as number[]));
-    }, [allStyles, breakpoints]);
+    let {
+      qa: defaultQa,
+      qaVal: defaultQaVal,
+      ...otherDefaultProps
+    } = defaultProps ?? {};
+
+    // eslint-disable-next-line react/display-name
+    _TastyComponent = forwardRef(
+      (allProps: AllBasePropsWithMods<K> & WithVariant<V>, ref) => {
+        let {
+          as,
+          styles,
+          variant,
+          breakpoints,
+          mods,
+          element,
+          qa,
+          qaVal,
+          ...otherProps
+        } = allProps;
+
+        let propStyles: Styles | null = (
+          (styleProps
+            ? (styleProps as StyleList).concat(BASE_STYLES)
+            : BASE_STYLES) as StyleList
+        ).reduce((map, prop) => {
+          if (prop in otherProps) {
+            map[prop] = otherProps[prop];
+
+            delete otherProps[prop];
+          }
 
-    if (mods) {
-      Object.assign(otherProps, modAttrs(mods));
-    }
+          return map;
+        }, {});
 
-    return (
-      <Element
-        as={as ?? originalAs}
-        data-element={element || defaultElement}
-        data-qa={qa || defaultQa}
-        data-qaval={qaVal || defaultQaVal}
-        {...otherDefaultProps}
-        {...otherProps}
-        ref={ref}
-        css={renderedStyles}
-      />
+        if (Object.keys(propStyles).length === 0) {
+          propStyles = null;
+        }
+
+        if (!styles || (styles && Object.keys(styles).length === 0)) {
+          styles = undefined;
+        }
+
+        const propStylesCacheKey = JSON.stringify(propStyles);
+        const stylesCacheKey = useMemo(() => {
+          return JSON.stringify(styles);
+        }, [styles]);
```

Comment:
I think it doesn't make sense to memoize styles here. We always pass styles object as is, without any memoization

Reply:
Sometimes we pass it as a variable (for example, while wrapping). Sometimes we pass an empty object. It will help in that cases.

Reply:
And will add overhead in all the others.